gtk/favicon.cc

```diff
@@ -1,172 +1,157 @@
 {
-    char* filename = favicon_get_cache_filename(host);
-    g_file_set_contents(filename, static_cast<gchar const*>(data), len, nullptr);
-    g_free(filename);
+    Glib::file_set_contents(favicon_get_cache_filename(host), data);
 }
 
-static GdkPixbuf* favicon_load_from_cache(char const* host)
+Glib::RefPtr<Gdk::Pixbuf> favicon_load_from_cache(std::string const& host)
 {
-    char* filename = favicon_get_cache_filename(host);
-    GdkPixbuf* pixbuf = gdk_pixbuf_new_from_file_at_size(filename, 16, 16, nullptr);
+    auto const filename = favicon_get_cache_filename(host);
 
-    if (pixbuf == nullptr) /* bad file */
+    try
     {
-        g_remove(filename);
+        return Gdk::Pixbuf::create_from_file(filename, 16, 16, false);
+    }
+    catch (Glib::Error const&)
+    {
+        g_remove(filename.c_str());
+        return {};
     }
-
-    g_free(filename);
-    return pixbuf;
 }
 
-static void favicon_web_done_cb(tr_session*, bool, bool, long, void const*, size_t, void*);
+void favicon_web_done_cb(tr_session*, bool, bool, long, void const*, size_t, void*);
 
-static gboolean favicon_web_done_idle_cb(gpointer vfav)
+bool favicon_web_done_idle_cb(favicon_data* fav)
 {
-    GdkPixbuf* pixbuf = nullptr;
-    gboolean finished = FALSE;
-    auto* fav = static_cast<favicon_data*>(vfav);
+    Glib::RefPtr<Gdk::Pixbuf> pixbuf;
+    bool finished = false;
 
-    if (fav->len > 0) /* we got something... try to make a pixbuf from it */
+    if (!fav->contents.empty()) /* we got something... try to make a pixbuf from it */
     {
-        favicon_save_to_cache(fav->host, fav->contents, fav->len);
+        favicon_save_to_cache(fav->host, fav->contents);
         pixbuf = favicon_load_from_cache(fav->host);
         finished = pixbuf != nullptr;
     }
 
     if (!finished) /* no pixbuf yet... */
     {
-        if (++fav->type == IMAGE_TYPES) /* failure */
+        if (++fav->type == image_types.size()) /* failure */
         {
-            finished = TRUE;
+            finished = true;
         }
         else /* keep trying */
         {
-            char* url = get_url(fav->host, fav->type);
-
-            g_free(fav->contents);
-            fav->contents = nullptr;
-            fav->len = 0;
-
-            tr_webRun(fav->session, url, favicon_web_done_cb, fav);
-            g_free(url);
+            fav->contents.clear();
+            tr_webRun(fav->session, get_url(fav->host, fav->type).c_str(), favicon_web_done_cb, fav);
         }
     }
 
     if (finished)
     {
-        fav->func(pixbuf, fav->data);
-        g_free(fav->host);
-        g_free(fav->contents);
-        g_free(fav);
+        fav->func(pixbuf);
+        delete fav;
     }
 
-    return G_SOURCE_REMOVE;
+    return false;
 }
 
-static void favicon_web_done_cb(
-    tr_session* session,
-    bool did_connect,
-    bool did_timeout,
-    long code,
+void favicon_web_done_cb(
+    tr_session* /*session*/,
+    bool /*did_connect*/,
+    bool /*did_timeout*/,
+    long /*code*/,
```

Comment:
[[maybe_unused]]

Reply:
This is a "definitely unused" type of situation, using the attribute seems misleading. I think it's suitable when you have a code with `#if`s that may or may not result in a parameter/variable being referenced, but if that's not the only case we're using it I'm okay with adding it...